colcon_sanitizer_reports/sanitizer_log_parser.py

```diff
@@ -0,0 +1,195 @@
+from collections import defaultdict
+import csv
+from io import StringIO
+import re
+from typing import Dict, List, Optional, Tuple
+
+
+class _KeysFactory:
+    """Finds key stack trace lines in sanitizer error subsections."""
+
+    @staticmethod
+    def _make_masked_line(line: str) -> str:
+        """Masks numbers in a sanitizer error line so that keys from similar errors will match."""
+        masked_line = line
+        masked_line = re.sub(r'0x[\dabcdef]+', '0xX', masked_line)
+        masked_line = re.sub(r' \d+ ', ' X ', masked_line)
+        masked_line = re.sub(r'#\d+', '#X', masked_line)
+
+        return masked_line
+
+    @staticmethod
+    def _make_keys_base(*, lines: Tuple[str], stack_begin_regexes: Tuple[str, ...]) -> Tuple[str]:
+        keys: List[str] = []
+
+        lines = iter(lines)
+        for stack_begin_regex in stack_begin_regexes:
+            for line in lines:
+                match = re.match(
+                    stack_begin_regex,
+                    _KeysFactory._make_masked_line(line)
+                )
+                if match is not None:
+                    break
+
+            for line in lines:
+                match = re.match(
+                    r'^\s*#X (0xX in|)\s*(?P<key>.*/ros2.*)\s*$',
```

Comment:
Let's make this a global constant e.g. `_FOO_REGEXP_STRING` or something and let's thoroughly document how to read this. For instance:
```py
# This regexp is used for ....
# Example of matching string: "sdjksldfjks"
_FOO_REGEXP_STRING = (
    r'^\s*'  # From the beginning of the line do X
    r'(0xX in|)'  # may contain xyz
)
```